ast_tools/passes/util.py

```diff
@@ -1,14 +1,112 @@
-import inspect
+from abc import ABCMeta, abstractmethod
 import ast
+import copy
+
+import inspect
 import typing as tp
 
+import libcst as cst
+
 from . import Pass
 from . import PASS_ARGS_T
 
 from ast_tools.stack import get_symbol_table, SymbolTable
-from ast_tools.common import get_ast, exec_def_in_file
+from ast_tools.common import get_ast, get_cst, exec_def_in_file, exec_str_in_file
+
+__ALL__ = ['begin_rewrite', 'end_rewrite', 'apply_ast_passes']
+
+
+def _issubclass(t, s) -> bool:
+    try:
+        return issubclass(t, s)
+    except TypeError:
+        return False
+
+
+class _DecoratorStripper(metaclass=ABCMeta):
+    @staticmethod
+    @abstractmethod
+    def get_decorators(tree): pass
+
+    @staticmethod
+    @abstractmethod
+    def set_decorators(tree, decorators): pass
+
+    @staticmethod
+    @abstractmethod
+    def get_name(node): pass
+
+    @classmethod
+    def strip(cls, tree, env, start_sentinel, end_sentinel):
+        decorators = []
+        in_group = False
+        first_group = True
+        # filter passes from the decorator list
+        for node in reversed(cls.get_decorators(tree)):
+            if not first_group:
+                decorators.append(node)
+                continue
+
+            name = cls.get_name(node)
+            deco = env[name]
+            if in_group:
+                if _issubclass(deco, end_sentinel):
+                    in_group = False
+                    first_group = False
+            elif start_sentinel is None or _issubclass(deco, start_sentinel):
+                if start_sentinel is end_sentinel:
+                    # Just remove current decorator
+                    first_group = False
+                else:
+                    in_group = True
+            else:
+                decorators.append(node)
+
+        tree = cls.set_decorators(tree, reversed(decorators))
+        return tree
+
+
+class _ASTStripper(_DecoratorStripper):
+    @staticmethod
+    def get_decorators(tree):
+        return tree.decorator_list
+
+
+    @staticmethod
+    def set_decorators(tree, decorators):
+        tree = copy.deepcopy(tree)
+        tree.decorator_list = decorators
+        return tree
+
+
+    @staticmethod
+    def get_name(node):
+        if isinstance(node, ast.Call):
+            return node.func.id
```

Comment:
Should add the same assertion as line 105 (assumes the decorator is just a simple name object).

Probably don't need to support this yet, but other cases that might be not too complicated to support would be attributes (e.g. `@m.combinational`) and function calls (e.g. `@foo()`).

Reply:
`@foo()`  type decorators are supported (its this case)

Supporting attribute, gets ... a bit gnarly.  Instead of just looking up decorators by name in the env we would need to eval the expression in the env.  Simple attributes like `@module.decorator` should be easy enough to support (we just need to assume that getattr is idempotent which seems reasonable).  However, if you start mixing calls and attributes (e.g. `@generate(expr).decorator`) then this assumption becomes less reasonable.